source/elements/oneDAL/source/algorithms/decomposition/pca.rst

```diff
@@ -1,3 +1,447 @@
+
+
+   .. member:: std::int64_t component_count = 0
+
+      The number of components :math:`r` in the trained model.
+
+      Getter
+         | ``std::int64_t get_component_count() const``
+
+      Invariants
+         | :expr:`component_count == centroids.row_count`
+
+
+
+Training :expr:`onedal::train(...)`
+-----------------------------------
+Input
+~~~~~
+::
+
+   class train_input {
+   public:
+      train_input();
+      train_input(const table& data);
+
+      const table& get_data() const;
+
+      train_input& set_data(const table&);
+   };
+
+.. class:: train_input
+
+   .. function:: train_input()
+
+      Creates an input for the training operation with the default attribute
+      values.
+
+
+   .. function:: train_input(const table& data)
+
+      Creates an input for the training operation with the given :expr:`data`.
+
+
+   .. member:: table data = table()
+
+      :math:`n \times p` table with the data to be clustered, where each row
+      stores one sample.
+
+      Getter & Setter
+         | ``const table& get_data() const``
+         | ``train_input& set_data(const table&)``
+
+
+Result
+~~~~~~
+::
+
+   class train_result {
+   public:
+      train_result();
+
+      const model& get_model() const;
+      const table& get_means() const;
+      const table& get_variances() const;
+      const table& get_eigenvalues() const;
+   };
+
+.. class:: train_result
+
+   .. function:: train_result()
+
+      Creates a result of the training operation with the default attribute
+      values.
+
+
+   .. member:: pca::model model = pca::model()
+
+      The trained K-means model.
+
+      Getter
+         | ``const model& get_model() const``
+
+
+   .. member:: table means = table()
+
+      :math:`1 \times r` table that contains mean value for each feature.
+
+      Getter
+         | ``const table& get_means() const``
+
+
+   .. member:: table variances = table()
+
+      :math:`1 \times r` table that contains variance for each feature.
+
+      Getter
+         | ``const table& get_variances() const``
+
+
+   .. member:: table eigenvalues = table()
+
+      :math:`1 \times r` table that contains eigenvalue for each feature.
+
+      Getter
+         | ``const table& get_eigenvalues() const``
+
+
+Operation semantics
+~~~~~~~~~~~~~~~~~~~
+.. namespace:: onedal
+.. function:: template <typename Descriptor> \
+              pca::train_result train(const Descriptor& desc, \
+                                      const pca::train_input& input)
+
+   :tparam Descriptor: PCA algorithm descriptor :expr:`pca::desc`.
+
+   Preconditions
+      | :expr:`input.data.is_empty == false`
+      | :expr:`input.data.column_count >= desc.component_count`
+
+   Postconditions
+      | :expr:`result.means.row_count == 1`
+      | :expr:`result.means.column_count == desc.component_count`
+      | :expr:`result.variances.row_count == 1`
+      | :expr:`result.variances.column_count == desc.component_count`
```

Comment:
those post-conditions for say variances: they should be >=0. Related to the topic of error handling and plausibility checks- what we want in dal